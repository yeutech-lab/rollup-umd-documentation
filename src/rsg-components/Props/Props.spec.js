/* eslint-disable react/prop-types */
import React from 'react';
import { parse } from 'react-docgen';
import PropsRenderer, { columns, getRowKey } from './PropsRenderer';
import { unquote, getType, showSpaces } from './util';

const propsToArray = props => Object.keys(props).map(name => ({ ...props[name], name }));

// Test renderers with clean readable snapshot diffs
export default function ColumnsRenderer({ props }) {
	return (
		<ul>
			{props.map((row, rowIdx) => (
				<li key={rowIdx}>
					{columns.map(({ render }, colIdx) => <div key={colIdx}>{render(row)}</div>)}
				</li>
			))}
		</ul>
	);
}

function render(propTypes, defaultProps = []) {
	const props = parse(`
		import { Component } from 'react';
		import PropTypes from 'prop-types';
		export default class Cmpnt extends Component {
			static propTypes = {
				${propTypes.join(',')}
			}
			static defaultProps = {
				${defaultProps.join(',')}
			}
			render() {
			}
		}
	`);
	return shallow(<ColumnsRenderer props={propsToArray(props.props)} />);
}

function renderFlow(propsType, defaultProps = []) {
	const props = parse(`
	  // @flow
		import * as React from 'react';
		type Props = {
			${propsType.join(',')}
		};
		export default class Cmpnt extends React.Component<Props> {
			static defaultProps = {
				${defaultProps.join(',')}
			}
			render() {
			}
		}
	`);
	return shallow(<ColumnsRenderer props={propsToArray(props.props)} />);
}

describe('PropsRenderer', () => {
	it('should render a table', () => {
		const actual = shallow(
			<PropsRenderer
				props={[{ type: { name: 'string' }, required: false, description: '', name: 'color' }]}
			/>
		);

		expect(actual).toMatchSnapshot();
	});
});

describe('props columns', () => {
	it('should render PropTypes.string', () => {
		const actual = render(['color: PropTypes.string']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.string with a default value', () => {
		const actual = render(['color: PropTypes.string'], ['color: "pink"']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.string.isRequired', () => {
		const actual = render(['color: PropTypes.string.isRequired']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.arrayOf', () => {
		const actual = render(['colors: PropTypes.arrayOf(PropTypes.string)']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.arrayOf(PropTypes.shape)', () => {
		const actual = render([
			'foos: PropTypes.arrayOf(PropTypes.shape({bar: PropTypes.number, baz: PropTypes.any}))',
		]);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.instanceOf', () => {
		const actual = render(['num: PropTypes.instanceOf(Number)']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.shape', () => {
		const actual = render([
			'foo: PropTypes.shape({bar: PropTypes.number.isRequired, baz: PropTypes.any})',
		]);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.shape with formatted defaultProps', () => {
		const actual = render(
			[
				`
				foo: PropTypes.shape({
					bar: PropTypes.number.isRequired,
					baz: PropTypes.any,
				})
			`,
			],
			[
				`
				foo: {
					bar: 123, baz() {
						return 'foo';
					},
					bing() {
						return 'badaboom';
					},
					trotskij: () => 1935,
					qwarc: { si: 'señor', },
				}
			`,
			]
		);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.shape defaultProps, falling back to Object', () => {
		const actual = render(
			[
				`
				foo: PropTypes.shape({
					bar: PropTypes.number.isRequired,
					baz: PropTypes.any,
				})
			`,
			],
			[
				`
				foo: somethingThatDoesntExist
			`,
			]
		);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.shape with description', () => {
		const actual = render([
			`foo: PropTypes.shape({
			/**
			 * Number
			 */
			bar: PropTypes.number.isRequired,
			/** Any */
			baz: PropTypes.any
		})`,
		]);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.objectOf', () => {
		const actual = render(['colors: PropTypes.objectOf(PropTypes.string)']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.objectOf(PropTypes.shape)', () => {
		const actual = render([
			`colors: PropTypes.objectOf(
			PropTypes.shape({
				bar: PropTypes.number.isRequired,
				baz: PropTypes.any
			})
		)`,
		]);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.oneOf', () => {
		const actual = render(['size: PropTypes.oneOf(["small", "normal", "large"])']);

		expect(actual).toMatchSnapshot();
	});

	it('should render PropTypes.oneOfType', () => {
		const actual = render(['union: PropTypes.oneOfType([PropTypes.string, PropTypes.number])']);

		expect(actual).toMatchSnapshot();
	});

	it('should render description in Markdown', () => {
		const actual = render(['/**\n * Label\n */\ncolor: PropTypes.string']);

		expect(actual).toMatchSnapshot();
	});

	it('should render unknown proptype for a prop when a relevant proptype is not assigned', () => {
		const actual = render([], ['color: "pink"']);

		expect(actual).toMatchSnapshot();
	});

	it('should render function body in tooltip', () => {
		const actual = render(['fn: PropTypes.func'], ['fn: (e) => console.log(e)']);

		expect(actual).toMatchSnapshot();
	});

	it('should render function defaultValue as code when undefined', () => {
		const actual = render(['fn: PropTypes.func'], ['fn: undefined']);

		expect(actual).toMatchSnapshot();
	});

	it('should render function defaultValue as code when null', () => {
		const actual = render(['fn: PropTypes.func'], ['fn: null']);

		expect(actual).toMatchSnapshot();
	});

	it('should render arguments from JsDoc tags', () => {
		const props = [
			{
				name: 'size',
				type: {
					name: 'number',
				},
				required: false,
				description: 'Test description',
				tags: {
					arg: [
						{
							name: 'Foo',
							description: 'Converts foo to bar',
							type: { name: 'Array' },
						},
					],
					param: [
						{
							name: 'Bar',
						},
					],
				},
			},
		];
		const actual = shallow(<ColumnsRenderer props={props} />);

		expect(actual).toMatchSnapshot();
	});

	it('should render return from JsDoc tags', () => {
		const getProps = tag => [
			{
				name: 'size',
				type: {
					name: 'number',
				},
				required: false,
				description: 'Test description',
				tags: {
					[tag]: [
						{
							title: 'Foo',
							description: 'Returns foo from bar',
							type: { name: 'Array' },
						},
					],
				},
			},
		];

		const actualForReturn = shallow(<ColumnsRenderer props={getProps('return')} />);

		expect(actualForReturn).toMatchSnapshot();

		const actualForReturns = shallow(<ColumnsRenderer props={getProps('returns')} />);

		expect(actualForReturns).toMatchSnapshot();
	});

	it('should render name as deprecated when tag deprecated is present', () => {
		const props = [
			{
				name: 'size',
				type: {
					name: 'number',
				},
				required: false,
				description: 'Test description',
				tags: {
					deprecated: [
						{
							title: 'deprecated',
							description: 'Do not use.',
						},
					],
				},
			},
		];
		const actual = shallow(<ColumnsRenderer props={props} />);

		expect(actual).toMatchSnapshot();
	});

	it('should render type string', () => {
		const actual = renderFlow(['foo: string']);

		expect(actual).toMatchSnapshot();
	});

	it('should render optional type string', () => {
		const actual = renderFlow(['foo?: string']);

		expect(actual).toMatchSnapshot();
	});

	it('should render type string with a default value', () => {
		const actual = renderFlow(['foo?: string'], ['foo: "bar"']);

		expect(actual).toMatchSnapshot();
	});

	it('should render literal type', () => {
		const actual = renderFlow(['foo?: "bar"']);

		expect(actual).toMatchSnapshot();
	});

	it('should render object type with body in tooltip', () => {
		const actual = renderFlow(['foo: { bar: string }']);

		expect(actual).toMatchSnapshot();
	});

	it('should render function type with body in tooltip', () => {
		const actual = renderFlow(['foo: () => void']);

		expect(actual).toMatchSnapshot();
	});

	it('should render union type with body in tooltip', () => {
		const actual = renderFlow(['foo: "bar" | number']);

		expect(actual).toMatchSnapshot();
	});

	it('should render tuple type with body in tooltip', () => {
		const actual = renderFlow(['foo: ["bar", number]']);

		expect(actual).toMatchSnapshot();
	});

	it('should render custom class type', () => {
		const actual = renderFlow(['foo: React.ReactNode']);

		expect(actual).toMatchSnapshot();
	});
});

describe('unquote', () => {
	it('should remove double quotes around the string', () => {
		const result = unquote('"foo"');
		expect(result).toBe('foo');
	});

	it('should remove single quotes around the string', () => {
		const result = unquote("'foo'");
		expect(result).toBe('foo');
	});

	it('should not remove quotes in the middle of the string', () => {
		const result = unquote('foo"bar');
		expect(result).toBe('foo"bar');
	});
});

describe('getType', () => {
	it('should return .type or .flowType property', () => {
		const result = getType({
			type: 'foo',
			flowType: 'bar',
		});
		expect(result).toBe('bar');
	});
});

describe('showSpaces', () => {
	it('should replace leading and trailing spaces with a visible character', () => {
		const result = showSpaces(' pizza ');
		expect(result).toBe('␣pizza␣');
	});
});

describe('getRowKey', () => {
	it('should return type name', () => {
		const result = getRowKey({ name: 'number' });
		expect(result).toBe('number');
	});
});
