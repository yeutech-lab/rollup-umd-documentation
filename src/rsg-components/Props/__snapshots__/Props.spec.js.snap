// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropsRenderer should render a table 1`] = `
<Styled(Table)
  columns={
    Array [
      Object {
        "caption": "Prop name",
        "render": [Function],
      },
      Object {
        "caption": "Type",
        "render": [Function],
      },
      Object {
        "caption": "Default",
        "render": [Function],
      },
      Object {
        "caption": "Description",
        "render": [Function],
      },
    ]
  }
  getRowKey={[Function]}
  rows={
    Array [
      Object {
        "description": "",
        "name": "color",
        "required": false,
        "type": Object {
          "name": "string",
        },
      },
    ]
  }
/>
`;

exports[`props columns should render PropTypes.arrayOf 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        colors
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        string[]
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.arrayOf(PropTypes.shape) 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foos
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        shape[]
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <Styled(Para)>
          <div
            key="bar"
          >
            <Styled(Name)>
              bar
            </Styled(Name)>
            : 
            <Styled(Type)>
              number
            </Styled(Type)>
          </div>
          <div
            key="baz"
          >
            <Styled(Name)>
              baz
            </Styled(Name)>
            : 
            <Styled(Type)>
              any
            </Styled(Type)>
          </div>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.instanceOf 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        num
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        Number
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.objectOf 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        colors
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        {string}
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.objectOf(PropTypes.shape) 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        colors
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        {shape}
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <Styled(Para)>
          <div
            key="bar"
          >
            <Styled(Name)>
              bar
            </Styled(Name)>
            : 
            <Styled(Type)>
              number
            </Styled(Type)>
             — 
            <Styled(Text)
              color="light"
              size="small"
            >
              Required
            </Styled(Text)>
          </div>
          <div
            key="baz"
          >
            <Styled(Name)>
              baz
            </Styled(Name)>
            : 
            <Styled(Type)>
              any
            </Styled(Type)>
          </div>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.oneOf 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        size
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        enum
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <Styled(Para)>
          <span>
            One of:
             
            <Group
              inline={true}
              separator=", "
            >
              <Styled(Code)
                key="\\"small\\""
              >
                small
              </Styled(Code)>
              <Styled(Code)
                key="\\"normal\\""
              >
                normal
              </Styled(Code)>
              <Styled(Code)
                key="\\"large\\""
              >
                large
              </Styled(Code)>
            </Group>
          </span>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.oneOfType 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        union
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        union
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <Styled(Para)>
          <span>
            One of type:
             
            <Group
              inline={true}
              separator=", "
            >
              <Styled(Type)
                key="string-0"
              >
                string
              </Styled(Type)>
              <Styled(Type)
                key="number-1"
              >
                number
              </Styled(Type)>
            </Group>
          </span>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.shape 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        shape
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <Styled(Para)>
          <div
            key="bar"
          >
            <Styled(Name)>
              bar
            </Styled(Name)>
            : 
            <Styled(Type)>
              number
            </Styled(Type)>
             — 
            <Styled(Text)
              color="light"
              size="small"
            >
              Required
            </Styled(Text)>
          </div>
          <div
            key="baz"
          >
            <Styled(Name)>
              baz
            </Styled(Name)>
            : 
            <Styled(Type)>
              any
            </Styled(Type)>
          </div>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.shape defaultProps, falling back to Object 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        shape
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Text)
        color="light"
        size="small"
        title="somethingThatDoesntExist"
        underlined={true}
      >
        Shape
      </Styled(Text)>
    </div>
    <div
      key="3"
    >
      <div>
        <Styled(Para)>
          <div
            key="bar"
          >
            <Styled(Name)>
              bar
            </Styled(Name)>
            : 
            <Styled(Type)>
              number
            </Styled(Type)>
             — 
            <Styled(Text)
              color="light"
              size="small"
            >
              Required
            </Styled(Text)>
          </div>
          <div
            key="baz"
          >
            <Styled(Name)>
              baz
            </Styled(Name)>
            : 
            <Styled(Type)>
              any
            </Styled(Type)>
          </div>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.shape with description 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        shape
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <Styled(Para)>
          <div
            key="bar"
          >
            <Styled(Name)>
              bar
            </Styled(Name)>
            : 
            <Styled(Type)>
              number
            </Styled(Type)>
             — 
            <Styled(Text)
              color="light"
              size="small"
            >
              Required
            </Styled(Text)>
             — 
            <Markdown
              inline={true}
              text="Number"
            />
          </div>
          <div
            key="baz"
          >
            <Styled(Name)>
              baz
            </Styled(Name)>
            : 
            <Styled(Type)>
              any
            </Styled(Type)>
             — 
            <Markdown
              inline={true}
              text="Any"
            />
          </div>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.shape with formatted defaultProps 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        shape
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Text)
        color="light"
        size="small"
        title="{
  bar: 123,
  baz: baz() {
          return 'foo';
      },
  bing: bing() {
          return 'badaboom';
      },
  trotskij: () => 1935,
  qwarc: {
    si: 'señor'
  }
}"
        underlined={true}
      >
        Shape
      </Styled(Text)>
    </div>
    <div
      key="3"
    >
      <div>
        <Styled(Para)>
          <div
            key="bar"
          >
            <Styled(Name)>
              bar
            </Styled(Name)>
            : 
            <Styled(Type)>
              number
            </Styled(Type)>
             — 
            <Styled(Text)
              color="light"
              size="small"
            >
              Required
            </Styled(Text)>
          </div>
          <div
            key="baz"
          >
            <Styled(Name)>
              baz
            </Styled(Name)>
            : 
            <Styled(Type)>
              any
            </Styled(Type)>
          </div>
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.string 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        color
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        string
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.string with a default value 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        color
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        string
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Code)>
        pink
      </Styled(Code)>
    </div>
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render PropTypes.string.isRequired 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        color
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        string
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Text)
        color="light"
        size="small"
      >
        Required
      </Styled(Text)>
    </div>
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render arguments from JsDoc tags 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        size
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        number
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <Markdown
          text="Test description"
        />
        <JsDoc
          arg={
            Array [
              Object {
                "description": "Converts foo to bar",
                "name": "Foo",
                "type": Object {
                  "name": "Array",
                },
              },
            ]
          }
          param={
            Array [
              Object {
                "name": "Bar",
              },
            ]
          }
        />
        <Styled(Arguments)
          args={
            Array [
              Object {
                "description": "Converts foo to bar",
                "name": "Foo",
                "type": Object {
                  "name": "Array",
                },
              },
              Object {
                "name": "Bar",
              },
            ]
          }
          heading={true}
        />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render custom class type 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        React.ReactNode
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Text)
        color="light"
        size="small"
      >
        Required
      </Styled(Text)>
    </div>
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render description in Markdown 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        color
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        string
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <Markdown
          text="Label"
        />
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render function body in tooltip 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        fn
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        func
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Text)
        color="light"
        size="small"
        title="(e) => console.log(e)"
        underlined={true}
      >
        Function
      </Styled(Text)>
    </div>
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render function defaultValue as code when null 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        fn
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        func
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Code)>
        null
      </Styled(Code)>
    </div>
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render function defaultValue as code when undefined 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        fn
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        func
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Code)>
        undefined
      </Styled(Code)>
    </div>
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render function type with body in tooltip 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        <Styled(Text)
          size="small"
          title="() => void"
          underlined={true}
        >
          <Styled(Type)>
            function
          </Styled(Type)>
        </Styled(Text)>
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Text)
        color="light"
        size="small"
      >
        Required
      </Styled(Text)>
    </div>
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render literal type 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        "bar"
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render name as deprecated when tag deprecated is present 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={true}
      >
        size
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        number
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <Markdown
          text="Test description"
        />
        <JsDoc
          deprecated={
            Array [
              Object {
                "description": "Do not use.",
                "title": "deprecated",
              },
            ]
          }
        />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render object type with body in tooltip 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        <Styled(Text)
          size="small"
          title="{ bar: string }"
          underlined={true}
        >
          <Styled(Type)>
            object
          </Styled(Type)>
        </Styled(Text)>
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Text)
        color="light"
        size="small"
      >
        Required
      </Styled(Text)>
    </div>
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render optional type string 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        string
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render return from JsDoc tags 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        size
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        number
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <Markdown
          text="Test description"
        />
        <JsDoc
          return={
            Array [
              Object {
                "description": "Returns foo from bar",
                "title": "Foo",
                "type": Object {
                  "name": "Array",
                },
              },
            ]
          }
        />
        <Styled(Argument)
          description="Returns foo from bar"
          returns={true}
          title="Foo"
          type={
            Object {
              "name": "Array",
            }
          }
        />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render return from JsDoc tags 2`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        size
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        number
      </Styled(Type)>
    </div>
    <div
      key="2"
    />
    <div
      key="3"
    >
      <div>
        <Markdown
          text="Test description"
        />
        <JsDoc
          returns={
            Array [
              Object {
                "description": "Returns foo from bar",
                "title": "Foo",
                "type": Object {
                  "name": "Array",
                },
              },
            ]
          }
        />
        <Styled(Argument)
          description="Returns foo from bar"
          returns={true}
          title="Foo"
          type={
            Object {
              "name": "Array",
            }
          }
        />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render tuple type with body in tooltip 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        <Styled(Text)
          size="small"
          title="[\\"bar\\", number]"
          underlined={true}
        >
          <Styled(Type)>
            tuple
          </Styled(Type)>
        </Styled(Text)>
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Text)
        color="light"
        size="small"
      >
        Required
      </Styled(Text)>
    </div>
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render type string 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        string
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Text)
        color="light"
        size="small"
      >
        Required
      </Styled(Text)>
    </div>
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render type string with a default value 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        string
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Code)>
        bar
      </Styled(Code)>
    </div>
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render union type with body in tooltip 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        foo
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        <Styled(Text)
          size="small"
          title="\\"bar\\" | number"
          underlined={true}
        >
          <Styled(Type)>
            union
          </Styled(Type)>
        </Styled(Text)>
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Text)
        color="light"
        size="small"
      >
        Required
      </Styled(Text)>
    </div>
    <div
      key="3"
    >
      <div>
        <Styled(Para)>
          <span />
        </Styled(Para)>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`props columns should render unknown proptype for a prop when a relevant proptype is not assigned 1`] = `
<ul>
  <li
    key="0"
  >
    <div
      key="0"
    >
      <Styled(Name)
        deprecated={false}
      >
        color
      </Styled(Name)>
    </div>
    <div
      key="1"
    >
      <Styled(Type)>
        unknown
      </Styled(Type)>
    </div>
    <div
      key="2"
    >
      <Styled(Code)>
        pink
      </Styled(Code)>
    </div>
    <div
      key="3"
    >
      <div>
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;
